modules/xfd.ts

```diff
@@ -0,0 +1,2218 @@
+		});
+	}
+
+	// Used as the label for the fieldset in the UI, and in the default notification edit summary
+	abstract getFieldsetLabel()
+
+	postRender(renderedFieldset: HTMLFieldSetElement) {}
+
+	getVenueWarning(): string | void {}
+
+	// Overridden for tfd, cfd, cfds
+	/**
+	 * Pre-process parameters, called from evaluate() and preview().
+	 */
+	preprocessParams(): void {}
+
+	// Overridden for ffd and rfd, which need special treatment
+	preview(form: HTMLFormElement) {
+		this.params = Morebits.quickForm.getInputData(form);
+		this.preprocessParams();
+		this.showPreview(form);
+	}
+
+	// This is good enough to use without override for all venues except rm
+	showPreview(form: HTMLFormElement) {
+		let templatetext = this.getDiscussionWikitext();
+		form.previewer.beginRender(templatetext, 'WP:TW'); // Force wikitext
+	}
+
+	abstract getDiscussionWikitext(): string
+
+	evaluate(): any {
+		this.params = Morebits.quickForm.getInputData(this.result);
+		this.preprocessParams();
+		Morebits.simpleWindow.setButtonsEnabled(false);
+		Morebits.status.init(this.result);
+
+		// XXX: Copied from the original twinklexfd, but this isn't great.
+		// Errors in tasks that execute early but don't affect creation of discussion page
+		// such as fetchCreatorInfo() should not print trigger this.
+		// But fetchCreatorInfo() is unlikely to error out so we're good for now.
+		Xfd.currentRationale = this.params.reason;
+		Morebits.status.onError(() => {
+			if (Xfd.currentRationale) {
+				Morebits.status.printUserText(Xfd.currentRationale, 'Your deletion rationale is provided below, which you can copy and paste into a new XFD dialog if you wish to try again:');
+				// only need to print the rationale once
+				Xfd.currentRationale = null;
+			}
+		});
+
+		return new Morebits.taskManager(this);
+	}
+
+	/**
+	 * Callback to redirect to the discussion page when everything is done. Relies on the discussion page
+	 * being known as either `this.params.discussionpage` or `this.params.logpage`.
+	 */
+	redirectToDiscussion() {
+		let redirPage = this.params.discussionpage || this.params.logpage;
+		Morebits.status.actionCompleted("Nomination complete, now redirecting to the discussion page");
+		setTimeout(() => {
+			window.location.href = mw.util.getUrl(redirPage);
+		}, Morebits.wiki.actionCompleted.timeOut);
+	}
+
+	autoEditRequest(pageobj: Morebits.wiki.page) {
+		let params = this.params;
+		let def = $.Deferred();
+		var talkName = new mw.Title(pageobj.getPageName()).getTalkPage().toText();
+		if (talkName === pageobj.getPageName()) {
+			pageobj.getStatusElement().error('Page protected and nowhere to add an edit request, aborting');
+			return def.reject();
+		} else {
+			pageobj.getStatusElement().warn('Page protected, requesting edit');
+
+			var editRequest = '{{subst:Xfd edit protected|page=' + pageobj.getPageName() +
+				'|discussion=' + params.discussionpage + '|tag=<nowiki>' + params.tagText + '\u003C/nowiki>}}'; // U+003C: <
```

Comment:
same comment as from the javascript about documenting why U+003C must be used manually

Reply:
This was added in #1091 though [on my suggestion](https://github.com/azatoth/twinkle/pull/1091#discussion_r507613528). In hindsight, yeah it was a stupid suggestion. Much better to do `'</no' + 'wiki>'`